scheduler/src/main/java/org/apache/falcon/notification/service/impl/SchedulerService.java

```diff
@@ -96,22 +94,26 @@ public void register(NotificationRequest notifRequest) throws NotificationServic
         if (request.getInstance() == null) {
             throw new NotificationServiceException("Request must contain an instance.");
         }
-        // When the instance is getting rescheduled for run. As in the case of suspend and resume.
-        Object obj = instancesToIgnore.getIfPresent(request.getInstance().getId());
-        if (obj != null) {
-            instancesToIgnore.invalidate(request.getInstance().getId());
-        }
         LOG.debug("Received request to schedule instance {} with sequence {}.", request.getInstance().getId(),
                 request.getInstance().getInstanceSequence());
         runQueue.execute(new InstanceRunner(request));
     }
 
     @Override
-    public void unregister(NotificationHandler handler, ID listenerID) {
+    public void unregister(NotificationHandler handler, ID listenerID) throws NotificationServiceException {
         // If ID is that of an entity, do nothing
         if (listenerID instanceof InstanceID) {
-            // Not efficient to iterate over elements to remove this. Add to ignore list.
-            instancesToIgnore.put((InstanceID) listenerID, new Object());
+            try {
+                InstanceID instanceID = (InstanceID) listenerID;
+                SortedMap<Integer, ExecutionInstance> instances = executorAwaitedInstances.get(instanceID
+                            .getEntityClusterID());
+                synchronized (instances) {
```

Comment:
Instances can be null or empty in case of no instances in ready. Null check and empty check is required here right ?